semantics/executable-spec/src/Control/State/Transition/Extended.hs

```diff
@@ -215,9 +229,32 @@ class
 class (STS sub, BaseM sub ~ BaseM super) => Embed sub super where
   -- | Wrap a predicate failure of the subsystem in a failure of the super-system.
   wrapFailed :: PredicateFailure sub -> PredicateFailure super
+  wrapEvent :: Event sub -> Event super
 
 instance STS sts => Embed sts sts where
   wrapFailed = id
+  wrapEvent = id
+
+data EventPolicy
+  = EventPolicyReturn
+  | EventPolicyDiscard
+
+data SingEP ep where
+  EPReturn  :: SingEP 'EventPolicyReturn
+  EPDiscard :: SingEP 'EventPolicyDiscard
+
+type family EventReturnType ep sts a :: Type where
+  EventReturnType 'EventPolicyReturn sts a = (a, [Event sts])
+  EventReturnType _ _ a = a
+
+type family EventConstraintType e sts m :: Constraint where
```

Comment:
ah, i don't it's necessary at all, then. it's always interpreted at https://github.com/input-output-hk/cardano-ledger-specs/pull/2244/files/0947eea86860dd4cbef1ed1e8a7f59587ae8c27c#diff-8d279cc8073de17e2b7c14f40f0bb530821125948ffa4d43c4712b9ae64d7261R514-R516 into the clause monad somehow. it would be trivial to make this interpret modifying some impure state, logging to stdout, whatever

Reply:
`EventReturnType` is gone, `EventConstraintType` is still useful

Reply:
Oh, wait, now I'm totally confused. I had thought the complete opposite way around...

Reply:
Namely: the point of the polymorphic return type is that consumers calling these methods (some outside of the ledger code) can continue to have exactly the same method signature, unless they choose to enable event logging. But the internal constraint seems to serve no useful purpose; that monad may just be fixed as the state monad, regardless of event logging. There's nothing the user can do to change it, so making it polymorphic seems unnecessary.

The initial idea behind the constraint polymorphism was to put the constraint on the base monad, thus allowing reporting into the base monad iff it supported such.

Reply:
ah ok, we do want to be polymorphic in the return type. nbd, I can just revert that change.

> The initial idea behind the constraint polymorphism was to put the constraint on the base monad, thus allowing reporting into the base monad if it supported such.

this is still desirable in my opinion, I can imagine at some point in the future introducing a new Event policy that modifies some impure state. I expect it will really only ever be used in `runClause`

Reply:
tbh after thinking about it a bit, having `EventReturnType` seems a bit unnecessary. I think since we need to pass the event return policy, there's already an API change so we may as well not have this type family in the return type